### PR TITLE
feat: link preview #520

### LINKED ISSUE #520: Url preview in chat message
Get operation:

1. returns null
2. Starts background fetch if null or stale (1d)
3. Updates content
4. Updates text entry (notifies consumers)

Dedup images by content sha

### PR DESCRIPTION
**Pull request by frolyo**

### CODE REVIEW

**Review by alexis-kochetov — APPROVED** [review]

File: src/dotnet/Chat.Service/Db/DbLinkPreview.cs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    public DbLinkPreview() { }
+    public DbLinkPreview(LinkPreview model) => UpdateFrom(model);
+
+    public LinkPreview ToModel()
+        => new() {
+            Id = new Symbol(Id),
+            Url = Url,
+            PreviewMediaId = new MediaId(ThumbnailMediaId),
```
Comment (alexyakunin) [inline src/dotnet/Chat.Service/Db/DbLinkPreview.cs:35]: Not the same name - intended?

File: src/dotnet/Chat.Service/Db/DbLinkPreview.cs [hunk 2]
```diff
@@ -0,0 +1,52 @@
+using System.ComponentModel.DataAnnotations;
+using System.ComponentModel.DataAnnotations.Schema;
+
+namespace ActualChat.Chat.Db;
+
+[Table("LinkPreviews")]
+public class DbLinkPreview : IHasId<string>, IRequirementTarget
```
Comment (alexyakunin) [inline src/dotnet/Chat.Service/Db/DbLinkPreview.cs:7]: Why no Version?

File: src/dotnet/Chat.UI.Blazor/Components/ChatView/Items/LinkPreview/link-preview.css [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+}
+.link-preview .c-title {
+    @apply text-xs md:text-base text-accent font-semibold;
+    @apply line-clamp-1 md:line-clamp-2;
+}
+.link-preview .c-description {
+    @apply text-xs md:text-base text-02;
+    @apply line-clamp-1 md:line-clamp-2;
+}
+
```
Comment (alexyakunin) [inline src/dotnet/Chat.UI.Blazor/Components/ChatView/Items/LinkPreview/link-preview.css:38]: Not sure why there is an extra empty line :)

File: src/dotnet/Chat.UI.Blazor/Services/ChatListUI.cs [hunk 4]
```diff
@@ -11,6 +11,7 @@ public partial class ChatListUI : WorkerBase, IHasServices, IComputeService, INo
     public static readonly int AllItemCountWhenLoading = 14;
     private static readonly TimeSpan MinNotificationInterval = TimeSpan.FromSeconds(5);
 
+    // modified implicitly after returned by GetItems
```
Comment (alexyakunin) [inline src/dotnet/Chat.UI.Blazor/Services/ChatListUI.cs:14]: Comments are like logged messages - sentence-like, but ok to omit the ending ".", coz they are typically a part of the "list".

File: src/dotnet/Chat/LinkPreview.cs [hunk 5]
```diff
@@ -0,0 +1,25 @@
+using MemoryPack;
+using Stl.Fusion.Blazor;
+
+namespace ActualChat.Chat;
+
+[ParameterComparer(typeof(ByValueParameterComparer))]
+[DataContract, MemoryPackable(GenerateType.VersionTolerant)]
+public sealed partial record LinkPreview : IHasId<Symbol>, IRequirementTarget
```
Comment (alexyakunin) [inline src/dotnet/Chat/LinkPreview.cs:8]: Needs IHasVersion<long>; ModifiedAt might be omitted, coz Version in our case is always ~ ModifiedAt (unless we want to distinguish between user-made and system-made modifications).

File: src/dotnet/Chat/LinkPreviewMode.cs [hunk 6]
```diff
@@ -0,0 +1,8 @@
+namespace ActualChat.Chat;
+
+public enum LinkPreviewMode
```
Comment (alexyakunin) [inline src/dotnet/Chat/LinkPreviewMode.cs:3]: `Dismiss` -> `None`.

Comment (alexyakunin) [reply src/dotnet/Chat/LinkPreviewMode.cs:3]: + It's better to indicate which enum value is the default one (zero).

**Pull request merged**